packages/chopsticks/src/plugins/trace-transaction/index.ts

```diff
@@ -0,0 +1,191 @@
+        },
+      })
+
+      console.log('decode extrinsic...')
+      const tx = meta.registry.createType('Extrinsic', hexToU8a(extrinsics[txIndex]))
+      console.dir({ extrinsic: tx.toHuman() }, { depth: null })
+
+      if (tx.method.section.toString() !== 'evm') {
+        throw new Error(`Unsupported extrinsic ${tx.method.toString()}`)
+      }
+
+      console.log('decode gas/storage limit...')
+      let gasLimit: BN
+      let storageLimit: BN
+      switch (tx.method.method.toString()) {
+        case 'call':
+        case 'create2':
+        case 'ethCall': {
+          gasLimit = tx.method.args[3] as any
+          storageLimit = tx.method.args[4] as any
+          break
+        }
+        case 'create': {
+          gasLimit = tx.method.args[2] as any
+          storageLimit = tx.method.args[3] as any
+          break
+        }
+        case 'ethCallV2': {
+          const GAS_MASK = new BN(100000)
+          const STORAGE_MASK = new BN(100)
+          const GAS_LIMIT_CHUNK = new BN(30000)
+          const MAX_GAS_LIMIT_CC = new BN(21) // log2(BLOCK_STORAGE_LIMIT)
+
+          const bbbcc = new BN((tx.method.args[4] as any).toBigInt()).mod(GAS_MASK)
+          const encodedGasLimit = bbbcc.div(STORAGE_MASK) // bbb
+          const encodedStorageLimit = bbbcc.mod(STORAGE_MASK) // cc
+
+          gasLimit = encodedGasLimit.mul(GAS_LIMIT_CHUNK)
+          storageLimit = new BN(2).pow(
+            encodedStorageLimit.gt(MAX_GAS_LIMIT_CC) ? MAX_GAS_LIMIT_CC : encodedStorageLimit,
+          )
+          break
+        }
+        default:
+          throw new Error(`Unsupported method ${tx.method.method.toString()}`)
+      }
+      console.log({ gasLimit: gasLimit.toString(), storageLimit: storageLimit.toString() })
+      console.log('building block...')
+      const { storageDiff } = await newBlock.call('Core_initialize_block', [header.toHex()])
+      newBlock.pushStorageLayer().setAll(storageDiff)
+      for (const extrinsic of extrinsics.slice(0, txIndex)) {
+        const { storageDiff } = await newBlock.call('BlockBuilder_apply_extrinsic', [extrinsic])
+        newBlock.pushStorageLayer().setAll(storageDiff)
+      }
+
+      console.log('running evm trace...')
+      const call = config.vm ? 'EVMTraceApi_trace_vm' : 'EVMTraceApi_trace_call'
+      const res = await newBlock.call(call, [
+        from,
+        to || '0x0000000000000000000000000000000000000000',
+        u8aToHex(meta.registry.createType('Vec<u8>', input).toU8a()),
+        u8aToHex(meta.registry.createType('Balance', hexToU8a(value)).toU8a()),
+        u8aToHex(meta.registry.createType('u64', gasLimit).toU8a()),
+        u8aToHex(meta.registry.createType('u32', storageLimit).toU8a()),
+        '0x00', // empty access list
+      ])
+
+      const logs = meta.registry
+        .createType<any>(`Result<Vec<${config.vm ? 'Step' : 'CallTrace'}>, DispatchError>`, res.result)
+        .asOk.toJSON()
+
+      const filepath = `${process.cwd()}/trace-${txHash}.json`
```

Comment:
generate random file is not good. should take path from user input.
the proper convention for cli tool is that the output are default to stdout, and logs are to stderr. in that way we can just `cmd > out.json`